File: Array/Easy/python/shiftZeros.py
def shiftZerostoEnd(arr):
    i=0
    while(i<len(arr)-1):
        if arr[i]==0:
            j=i+1
            while j<len(arr)-1 and arr[j] == 0:
                j+=1
            arr[i],arr[j] = arr[j],arr[i]
        i+=1

arr= [1,0,4,0,0,3,5,6,20,0]
shiftZerostoEnd(arr)
print(arr)


#optimal solution
'''
TC: O(N)
SC: O(1)
'''
arr= [1,0,4,0,0,3,5,6,20,0]
n = len(arr)
# O(X)
for i in range(n):
    if(arr[i] == 0):
        j=i
        break
#O(N-X)
for i in range(j+1,n):
    if(arr[i]!=0):
        arr[i],arr[j] = arr[j], arr[i]
        j+=1
print(arr)

File: Array/Easy/js/shiftZeros.js
/** Using extra space */
function shiftZerostoEnd(nums) {
    let nonZero = [];
    let zeros = 0;
    for(let i=0;i<nums.length;i++){
        if(nums[i]!=0){
            nonZero.push(nums[i])
        } else {
            zeros++;
        }
    }
    console.log(zeros)
    while(zeros > 0){
        nonZero.push(0)
        zeros--
    }
    return nonZero
};

/** Modifying the same array */
function shiftZerostoEnd2(arr){
    let i=0
    while (i<arr.length-1){
        if(arr[i] == 0){
            j=i+1
            while(j<arr.length-1 && arr[j]==0){
                j++
            }
            [arr[i],arr[j]] = [arr [j],arr[i]]                
        }
        i++
    }
}
arr= [1,0,4,0,0,3,5,6,20,0]
console.log('After Shifting zeros : ',shiftZerostoEnd(arr))
shiftZerostoEnd2(arr)
console.log('After Shifting zeros with optimal approach: ',arr)